src/Nethermind/Nethermind.Network.Discovery/Kademlia/Kademlia.cs

```diff
@@ -328,32 +381,49 @@ public async Task Bootstrap(CancellationToken token)
 
         // Refreshes all bucket. one by one. That is not empty.
         // A refresh means to do a k-nearest node lookup for a random hash for that particular bucket.
-        for (var i = 0; i < _buckets.Length; i++)
+        if (_useTreeImplementation)
         {
-            token.ThrowIfCancellationRequested();
+            for (var i = 0; i < Hash256XORUtils.MaxDistance + 1; i++)
+            {
+                token.ThrowIfCancellationRequested();
 
-            if (_buckets[i].Count > 0)
+                if (_bucketTree.GetAllAtDistance(i).Length > 0)
+                {
+                    ValueHash256 nodeToLookup = Hash256XORUtils.GetRandomHashAtDistance(_currentNodeIdAsHash, i);
+                    await LookupNodesClosest(nodeToLookup, token);
+                }
+            }
+        }
+        else
+        {
+            for (var i = 0; i < _buckets.Length; i++)
             {
-                ValueHash256 nodeToLookup = Hash256XORUtils.GetRandomHashAtDistance(_currentNodeIdAsHash, i);
-                await LookupNodesClosest(nodeToLookup, token);
+                token.ThrowIfCancellationRequested();
+
+                if (_buckets[i].Count > 0)
+                {
+                    ValueHash256 nodeToLookup = Hash256XORUtils.GetRandomHashAtDistance(_currentNodeIdAsHash, i);
+                    await LookupNodesClosest(nodeToLookup, token);
+                }
             }
         }
 
-        _logger.Info($"Boostrap completed. Took {sw}. Bucket sizes (from 230) {string.Join(",", _buckets[200..].Select((b) => b.Count).ToList())}");
+        _logger.Info($"Bootstrap completed. Took {sw}. Bucket sizes (from 230) {string.Join(",", Enumerable.Range(200, 56).Select(i => GetAllAtDistance(i).Length))}");
     }
 
     public IEnumerable<TNode> IterateNeighbour(ValueHash256 hash)
     {
         int startingDistance = Hash256XORUtils.CalculateDistance(_currentNodeIdAsHash, hash);
         foreach (var bucketToGet in EnumerateBucket(startingDistance))
```

Comment:
I *think* the enumerate bucket should be difference. Or this does not make a different at the moment.